Semaphore

82marbag opened the pull request:
*Issue #, if available:* #144

*Description of changes:*
* Fix an atomic operation issue where the asm code was invalid
* Add semaphore support with a similar interface as in posix semaphores

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

82marbag committed:
atomic: atomic_sub uses atomic_add

* `atomic_sub` uses `atomic_add` with a negative number
* `atomic_dec` takes an implicit `1` parameter

Signed-off-by: Daniele Ahmed <ahmeddan amazon c;0m >

wipawel commented on include/arch/x86/atomic.h:
```diff
@@ -151,28 +151,18 @@ static inline void atomic64_dec(atomic64_t *v) {
-                 :
-                 : "memory");
-    return val;
+    return atomic_add_return(v, -n);
```
I would keep original code here

82marbag replied:
The problem is xsub does not exist

wipawel commented on include/arch/x86/atomic.h:
```diff
@@ -151,28 +151,18 @@ static inline void atomic64_dec(atomic64_t *v) {
 
 static inline int64_t atomic64_sub_return(atomic64_t *v, int64_t n) {
-    int64_t val = n;
-    asm volatile("lock xsubq %[n], %[addr];"
-                 : [ n ] "+r"(val), [ addr ] "+m"(v->counter)
-                 :
-                 : "memory");
-    return val;
+    return atomic64_add_return(v, -n);
 }
```
I would keep original code here

wipawel commented on include/ktfsemaphore.h:
```diff
@@ -0,0 +1,44 @@
+/*
```
Please rename the file to semaphore.h

wipawel commented on include/ktfsemaphore.h:
```diff
@@ -0,0 +1,44 @@
+#define KTF_SEMAPHORE_H
+
+/*
+    TODO: rename to semaphore.h
+    When https://github.com/awslabs/ktf/issues/89 is fixed.
```
It is fixed now

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,90 @@
+#include <mm/slab.h>
+
+struct sem {
+    atomic64_t v;
```
atomic32_t

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,90 @@
+#include <atomic.h>
+#include <mm/slab.h>
+
+struct sem {
```
I would move it to semaphore.h and add a typedef

82marbag replied:
This is so that it's not possible to access the structure directly. If you feel this is too much for ktf, I'll move it as you suggest (in posix it's implemented similarly to this)

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,90 @@
+    if (value > _U32(-1) - 1) {
+        return false;
+    }
+    sem_t sem = kmalloc(sizeof(struct sem));
```
Please add dedicated sem_alloc() and make sem_init() being able to initialize also local variables

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,90 @@
+
+bool sem_init(sem_t *_sem, uint32_t value) {
+    /* Do not allow more than 2^32-1 threads to access this semaphore */
+    if (value > _U32(-1) - 1) {
```
Please add corresponding define to limits.h and use it here.

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,90 @@
+}
+
+bool sem_post(sem_t *_sem) {
+    sem_t sem = *_sem;
```
Not needed

82marbag replied:
The function or the declaration?

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,90 @@
+bool sem_post(sem_t *_sem) {
+    sem_t sem = *_sem;
+    atomic64_inc(&sem->v);
+    return true;
```
Not needed

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,90 @@
+        /* TODO: yield() when the scheduler supports it */
+        cpu_relax();
+    }
+    return true;
```
Not needed

82marbag replied:
Removed, it was to make it similar to the posix one

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,90 @@
+}
+
+bool sem_destroy(sem_t *_sem) {
+    sem_t sem = *_sem;
```
Not needed. Use just sem here and everywhere

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,90 @@
+    sem_t sem = kmalloc(sizeof(struct sem));
+
+    atomic_set(&sem->v, value);
+    sem->max_users = value;
```
Why do we need max_users?

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,90 @@
+bool sem_destroy(sem_t *_sem) {
+    sem_t sem = *_sem;
+    sem->max_users = 0;
+    atomic_set(&sem->v, 0);
```
I suppose you need to actually free the memory too?

82marbag replied:
Thanks

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,90 @@
+    sem_t sem = *_sem;
+    sem->max_users = 0;
+    atomic_set(&sem->v, 0);
+    return true;
```
Not needed.

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,90 @@
+}
+
+bool sem_trywait(sem_t *_sem) {
+    bool can_proceed = false;
```
Not needed. Can return directly.

wipawel commented on include/arch/x86/atomic.h:
```diff
@@ -158,9 +158,7 @@ static inline int64_t atomic64_sub_return(atomic64_t *v, int64_t n) {
-static inline int32_t atomic_dec_return(atomic_t *v) {
-    return atomic_sub_return(v, 1);
-}
+static inline int32_t atomic_dec_return(atomic_t *v) { return atomic_sub_return(v, 1); }
```
this should go to the other commit

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,80 @@
+
+bool sem_init(sem_t *sem, uint32_t value) {
+    /* Do not allow more than 2^31-1 threads to access this semaphore */
+    if (value > _U32(-1) / 2) {
+        return false;
+    }
```
this could be a BUG_ON, then this function does not have to return anything

wipawel commented on include/semaphore.h:
```diff
@@ -0,0 +1,45 @@
+
+struct sem {
+    atomic_t v;
+    uint32_t max_users;
```
What is this field for?

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,80 @@
+
+bool sem_getvalue(sem_t *sem, uint32_t *v) {
+    *v = atomic_read(&(sem->v));
+    return true;
```
Not needed

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,80 @@
+#include <mm/slab.h>
+#include <semaphore.h>
+
+sem_t *sem_alloc() { return kmalloc(sizeof(sem_t)); }
```
I don't think we need this. A macro initializer for global variables and the sem_init() below should be enough.

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,80 @@
+    return true;
+}
+
+void sem_destroy(sem_t *sem) {
+    sem->max_users = 0;
+    atomic_set(&(sem->v), 0);
+    kfree(sem);
+}
```
Probably not needed

82marbag committed:
concurrency: introduce semaphore support

Add semaphore support.
The interface is similar to posix's. This
implementation is lock-free and supports at most 2^31-1 concurrent
max users for each semaphore.

Signed-off-by: Daniele Ahmed <ahmeddan amazon c;0m >

wipawel commented on include/semaphore.h:
```diff
@@ -0,0 +1,43 @@
+};
+typedef struct sem sem_t;
+
+#define SEM_INIT(sem, value) sem_init(&sem, value)
```
This won't work outside of function scope

wipawel commented on lib/semaphore.c:
```diff
@@ -0,0 +1,58 @@
+
+    return false;
+}
+
```
What happened to sem_post()?

wipawel commented on include/semaphore.h:
```diff
@@ -0,0 +1,46 @@
+typedef struct sem sem_t;
+
+#define SEM_INIT()                                                                       \
+    {                                                                                    \
+        { 0 }                                                                            \
+    }
```
`#define SEM_INIT(value) { .v = (value) }`

82marbag replied:
This still doesn't check the bounds of `value` as in the init function, is it ok?

wipawel replied:
You could use `BUILD_BUG_ON()`, but it is also ok not to.

wipawel committed:
Merge branch 'mainline' into semaphore
